### PR TITLE
Add test ensuring that two surface integral methods are equivalent.

### PR DESCRIPTION
**Pull request by markscheel**
These two methods are derived by considering a conservation law
expressed using coordinate integrals (which explicitly include
factors of sqrt(g)) and considering the same conservation law
expressed using integrals over volume forms.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp [hunk 1]
```diff
@@ -319,6 +325,84 @@ void test_area(const Solution& solution, const Strahlkorper<Fr>& strahlkorper,
   CHECK(irreducible_mass == approx(expected_irreducible_mass));
 }
 
+// Let P = surface_integral_of_scalar(J^i \tilde{s}_i \sqrt(-g))
+// where \tilde{s}_i is the spatial unit one-form to the Strahlkorper,
+// normalized with the flat metric \tilde{s}_i \tilde{s}_j \delta^{ij} = 1,
+// and where euclidean_area_element is passed into surface_integral_of_scalar.
```
Comment (fmahebert) [inline tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp:333]: Define g (spacetime metric?), and perhaps (I don't feel strongly) also define J^i (a flux through the surface?)

File: tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp [hunk 2]
```diff
@@ -319,6 +325,84 @@ void test_area(const Solution& solution, const Strahlkorper<Fr>& strahlkorper,
...
+//
+// Let Q = surface_integral_of_scalar(J^i s_i \alpha), where \alpha is the
+// lapse, where s_i is the spatial unit one-form to the Strahlkorper,
+// normalized with the 3 metric s_i s_j \gamma^{ij} = 1, and where
+// the area_element computed with $\gamma_{ij}$ is passed into
+// surface_integral_of_scalar.
+//
+// This tests that P=Q for an arbitrary 3-vector J^i.
```
Comment (fmahebert) [inline tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp]: Very minor, feel free to ignore: the comment uses P vs. Q, but the code uses 1 vs. 2. Can these be made consistent? For example, by changing P to "I1" and Q to "I2", or whatever else you prefer.

File: tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp [hunk 3]
```diff
@@ -319,6 +325,84 @@ void test_area(const Solution& solution, const Strahlkorper<Fr>& strahlkorper,
...
+      jacobian, normal_one_form, radius, r_hat);
+
+  const auto normal_one_form_euclidean_magnitude =
+    dot_product(normal_one_form, normal_one_form);
+  const auto normal_one_form_magnitude =
+    dot_product(normal_one_form, normal_one_form, inverse_spatial_metric);
+
+  auto normal_one_form_flat = normal_one_form;
+  auto normal_one_form_curved = normal_one_form;
+  for(size_t i=0;i<3;++i) {
```
Comment (fmahebert) [inline tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp]: clang-format is needed here

File: tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp [hunk 4]
```diff
@@ -685,6 +769,22 @@ SPECTRE_TEST_CASE("Unit.ApparentHorizons.StrahlkorperGr.AreaElement",
             expected_area, expected_irreducible_mass);
 
   test_euclidean_area_element(kerr_horizon);
+
+  test_integral_correspondence(gr::Solutions::Minkowski<3>{},
+                               kerr_horizon);
+  test_integral_correspondence(gr::Solutions::KerrSchild{mass, spin, center},
+                               kerr_horizon);
+
+  // Check that this still works for a surface inside and outside the horizon
```
Comment (fmahebert) [inline tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp]: "this still works" -> "the two expressions giving the surface integral are still equal" or some rephrasing that increases clarity.

File: tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp [hunk 5]
```diff
@@ -685,6 +769,22 @@ SPECTRE_TEST_CASE("Unit.ApparentHorizons.StrahlkorperGr.AreaElement",
...
+      l_max, l_max, 0.9 * get(horizon_radius), center);
+  const auto outside_kerr_horizon = Strahlkorper<Frame::Inertial>(
+      l_max, l_max, 2.0 * get(horizon_radius), center);
+  test_integral_correspondence(gr::Solutions::KerrSchild{mass, spin, center},
+                               inside_kerr_horizon);
+  test_integral_correspondence(gr::Solutions::KerrSchild{mass, spin, center},
+                               outside_kerr_horizon);
 }
 
 SPECTRE_TEST_CASE("Unit.ApparentHorizons.StrahlkorperGr.SurfaceInteralOfScalar",
```
Comment (fmahebert) [inline tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp]: Consider adding a new commit that fixes the typo ("Interal") in this test name.

**Review by fmahebert — COMMENTED** [review]
Small suggestions on the documentation

File: tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp [hunk 6]
```diff
@@ -319,6 +325,84 @@ void test_area(const Solution& solution, const Strahlkorper<Fr>& strahlkorper,
...
 }
 
+// Let P = surface_integral_of_scalar(J^i \tilde{s}_i \sqrt(-g))
+// where \tilde{s}_i is the spatial unit one-form to the Strahlkorper,
+// normalized with the flat metric \tilde{s}_i \tilde{s}_j \delta^{ij} = 1,
+// and where euclidean_area_element is passed into surface_integral_of_scalar.
+//
+// Let Q = surface_integral_of_scalar(J^i s_i \alpha), where \alpha is the
+// lapse, where s_i is the spatial unit one-form to the Strahlkorper,
+// normalized with the 3 metric s_i s_j \gamma^{ij} = 1, and where
```
Comment (hrueter) [inline tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp]: spatial metric

File: tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp [hunk 7]
```diff
@@ -319,6 +325,84 @@ void test_area(const Solution& solution, const Strahlkorper<Fr>& strahlkorper,
...
+// the area_element computed with $\gamma_{ij}$ is passed into
+// surface_integral_of_scalar.
+//
+// This tests that P=Q for an arbitrary 3-vector J^i.
+template <typename Solution, typename Fr>
```
Comment (hrueter) [inline tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp]: `Fr` -> `Frame`

File: tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp [hunk 8]
```diff
@@ -319,6 +325,84 @@ void test_area(const Solution& solution, const Strahlkorper<Fr>& strahlkorper,
...
+  const auto euclidean_area_element = StrahlkorperGr::euclidean_area_element(
+      jacobian, normal_one_form, radius, r_hat);
+
+  const auto normal_one_form_euclidean_magnitude =
+    dot_product(normal_one_form, normal_one_form);
+  const auto normal_one_form_magnitude =
+    dot_product(normal_one_form, normal_one_form, inverse_spatial_metric);
+
+  auto normal_one_form_flat = normal_one_form;
+  auto normal_one_form_curved = normal_one_form;
```
Comment (hrueter) [inline tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp]: `unit_normal_one_form_flat` and `unit_normal_one_form_curved`

File: tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp [hunk 9]
```diff
@@ -319,6 +325,84 @@ void test_area(const Solution& solution, const Strahlkorper<Fr>& strahlkorper,
...
+  for(size_t i=0;i<3;++i) {
+    normal_one_form_flat.get(i) /=
+        sqrt(get<>(normal_one_form_euclidean_magnitude));
```
Comment (hrueter) [inline tests/Unit/ApparentHorizons/Test_StrahlkorperGr.cpp]: `get` without `<>`

**Review by hrueter — CHANGES_REQUESTED** [review]
group review by @hrueter (main) , @nilsleiffischer

**Comment by fmahebert** [comment]
LGTM, I am happy for a squash once @hrueter is ready.

**Comment by kidder** [comment]
@hrueter can you confirm if you are happy with the fixup

**Review by hrueter — DISMISSED** [review]

**Comment by hrueter** [comment]
We are happy with the fixup.

**Review by kidder — COMMENTED** [review]
@markscheel I think this just needs to be squashed and it can be merged...

**Review by hrueter — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**